Adapts selenium tests to che multiuser changes

akorneta opened the pull request:
### What does this PR do?
Adapts selenium tests to che multiuser changes.
Adds keycloak auth client to allow get and refresh access tokens for test users.
Adds tests startup option:
`--multiuser`
Note that without this parameter, selenium tests will proceed the same flow as on single user che.

Adds configuration properties:
Defines default test user for che single / multiuser selenium tests
`che.test_user.email`
`che.test_user.password`

Defines admin test user for che multiuser selenium tests
`che.admin_user.email`
`che.admin_user.password`

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/5930

tolusha commented:
Is ready to review?

akorneta commented:
@tolusha yes, in general, things related to che multiuser almost done except the additional user creation.
Another part that related to launch che single user will be done in this pr asap.

dmytro-ndp commented on selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestAdminHttpJsonRequestFactory.java:
```diff
@@ -10,21 +10,32 @@
-    this.adminProvider = adminProvider;
+  public TestAdminHttpJsonRequestFactory(
+      TestAuthServiceClient authServiceClient,
+      @Named("che.admin_user.email") String adminEmail,
```
It's Che-specific settings which should be re-written in tests for Codenvy onprem/saas. 
I propose to create **TestAdminEmailProvider** and **TestAdminPasswordProvider** interfaces and next implementations for them:
- **CheTestAdminEmailProvider** and **CheTestAdminPasswordProvider** implementations in selenium tests for Che (where che.admin_user.email and che.admin_user.password" are using);
- **CodenvyTestAdminEmailProvider** and **CodenvyTestAdminPasswordProvider** implementations for Codenvy (where codenvy.admin.name, codenvy.admin.initial_password, and codenvy.admin.password similarly to the class https://github.com/codenvy/codenvy/blob/master/selenium/codenvy-selenium-test/src/main/java/com/codenvy/selenium/core/user/OnpremAdminTestUser.java)

dmytro-ndp commented on selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestAdminHttpJsonRequestFactory.java:
```diff
@@ -10,21 +10,32 @@
+    try {
+      return authServiceClient.login(adminEmail, adminPassword);
+    } catch (Exception ex) {
+      throw new RuntimeException("Failed to get access token for admin user");
```
Let's log admin's email here as well so as it could be different.

akorneta replied:
added

dmytro-ndp commented on selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestDefaultUserHttpJsonRequestFactory.java:
```diff
@@ -10,22 +10,37 @@
+      return authServiceClient.login(testUser.getEmail(), testUser.getPassword());
+    } catch (Exception ex) {
+      throw new RuntimeException(
+          format("Failed to access get token for user '%s'", testUser.getName()));
```
It's better to log user's email for the full information so as user's name is usual just the part of email before "@"

dmytro-ndp commented on selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestDefaultUserHttpJsonRequestFactory.java:
```diff
@@ -10,22 +10,37 @@
   protected String getAuthToken() {
-    return testUserProvider.get().getAuthToken();
+    try {
+      return authServiceClient.login(testUser.getEmail(), testUser.getPassword());
```
we can use **testUser.getAuthToken()** here, can't we?

dmytro-ndp commented on selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestDefaultUserHttpJsonRequestFactory.java:
```diff
@@ -10,22 +10,37 @@
+ * @author Anton Korneta
+ */
 @Singleton
 public class TestDefaultUserHttpJsonRequestFactory extends TestHttpJsonRequestFactory {
```
This class looks like redundant after default test user is replaced by regular test user, doesn't it?

akorneta replied:
yes you are right

dmytro-ndp commented on selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestUserHttpJsonRequestFactory.java:
```diff
@@ -10,22 +10,33 @@
   protected String getAuthToken() {
-    return this.authToken;
+    try {
+      return authServiceClient.login(testUser.getEmail(), testUser.getPassword());
```
we can use **testUser.getAuthToken()** here, can't we?

akorneta replied:
Well, we need to guarantee that returned access token is valid and not expired so it's better to use authServiceClient.login call to get/refresh/login user's token

dmytro-ndp replied:
But **testUser.getAuthToken()** can refresh user's token every time we call it, can't it?

dmytro-ndp commented on selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestUserHttpJsonRequestFactory.java:
```diff
@@ -10,22 +10,33 @@
+      return authServiceClient.login(testUser.getEmail(), testUser.getPassword());
+    } catch (Exception ex) {
+      throw new RuntimeException(
+          format("Failed to access get token for user '%s'", testUser.getName()));
```
It's better to log user's email for the full information so as user's name is usual just the part of email before "@"

dmytro-ndp commented on selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/user/TestUserFactory.java:
```diff
@@ -17,7 +17,13 @@
 
-  TestUser create(@Assisted("email") String email, @Assisted("password") String password);
+  /** Creates new test user with generated e-mail and password */
+  TestUserImpl create();
```
What is the propose of this method? 
It looks as redundant so as simple `@Inject TestUser test user;` could be used instead of testUserFactory.create()

akorneta replied:
This method needs in case when in tests there need to be some user with generated email and password. So for codenvy I planed to bind the Test user to implementation wich call this method.

dmytro-ndp replied:
I see. But we don't need assisted injection, that is TestUserFactory, for creation an instance from constructor without **@Assisted** parameters at all. 
@Inject annotation is able to provide required instance of TestUser.

akorneta replied:
removed method + changed annotation

dmytro-ndp commented on selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/user/TestUserImpl.java:
```diff
@@ -45,17 +44,32 @@
 
   /** To instantiate user with generated email and password. */
-  @Inject
+  @AssistedInject
```
Did you test it? 
It will lead to an error of injection when we try to create random user by the command `@inject TestUser testUser;`.

akorneta replied:
yes for usecase described [here](https://github.com/eclipse/che/pull/6456#discussion_r141295205) it works pretty fine

dmytro-ndp commented on selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/core/TestUserServiceClientProvider.java:
```diff
@@ -0,0 +1,37 @@
+
+  @Override
+  public TestUserServiceClient get() {
+    return isMultiUser ? cheTestUserServiceClient : cheTestUserServiceClient;
```
Strange useless command. IMHO it's better to explicitly define proper user service clients in **CheSeleniumSingleUserModule** and **CheSeleniumMultiUserModule**.

dmytro-ndp commented on selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/client/TestUserServiceClient.java:
```diff
@@ -10,65 +10,69 @@
+ */
+public interface TestUserServiceClient {
 
-  public User getByEmail(String email) throws Exception {
```
We need this method for https://github.com/codenvy/codenvy/blob/master/selenium/codenvy-selenium-test/src/test/java/com/codenvy/selenium/factory/AuthenticateAndAcceptFactoryThroughGitHubOAuthTest.java

akorneta replied:
I'll adapt this code in codenvy PR

dmytro-ndp commented on selenium/che-selenium-test/src/test/resources/conf/selenium.properties:
```diff
@@ -24,3 +24,21 @@ github.auxiliary.password=
+che.test_user.password=
+
+# Defines additional test user
+che.aditional_test_user.email=
```
I can't find the usage of this properties. Can you, please, write an example of usage?

akorneta replied:
I'm working on it and will notify when it will be done

akorneta replied:
@dmytro-ndp  I deleted these properties because for the first step of integration selenium tests with  che multiuser it's overhead.

dmytro-ndp replied:
I see. We can add it in future on demand.

dmytro-ndp commented on selenium/che-selenium-test/src/test/resources/suites/CheSuite.xml:
```diff
@@ -886,18 +886,6 @@
                 </methods>
             </class>
             <class name="org.eclipse.che.selenium.refactor.types.TestFailTest"/>
-            <class name="org.eclipse.che.selenium.subversion.CommitTest"/>
```
It have not been finally decided to remove support of SVN plugin yet https://github.com/eclipse/che/issues/5484#issuecomment-328075154

tolusha replied:
@akorneta Pls, restore removed tests.

dmytro-ndp commented on selenium/che-selenium-test/src/test/resources/suites/PassingSuite.xml:
```diff
@@ -0,0 +1,775 @@
+<!--
```
What is this suite for? We have already have CheSuite.xml, and it is using by default.

akorneta replied:
removed

dmytro-ndp commented on selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/CheAdminHttpJsonRequestFactory.java:
```diff
@@ -10,34 +10,18 @@
-  private final String adminEmail;
-  private final String adminPassword;
-  private final TestAuthServiceClient authServiceClient;
+public class CheAdminHttpJsonRequestFactory extends TestUserHttpJsonRequestFactory {
```
Usually we add **Test** word into the name of selenium test framework classes to avoid confusion with Che product codebase.

akorneta commented:
Fixed single user che test launching, for codenvy and saas assemblies there will be separate prs. Removed additional che user properties.
@tolusha @dmytro-ndp please take a look.

tolusha commented on selenium/che-selenium-test/selenium-tests.sh:
```diff
@@ -14,7 +14,24 @@ export CALLER=$(basename $0)
+
+export CHE_MULTIUSER=$cheMultiuserValue
+
+set -- "${params[@]}"
```
Make it simpler. Just export what you need.

dmytro-ndp replied:
We should remove **--multiuser** parameter from list of parameters because of **webdriver.sh** doesn't recognize it and throws an error.

tolusha replied:
That's what I told. In case of multi use assembly:
```
export CHE_MULTIUSER=true
```

And then check environment variable in the code.

dmytro-ndp replied:
We should remove _--multiuser_ parameter from parameters because otherwise it will be been passed to **webdriver.sh** as it and will cause an error at start of execution of selenium tests:
> (target/bin/webdriver.sh "$TESTS_SCOPE" **$@**)

tolusha commented on selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/core/factory/TestFactory.java:
```diff
@@ -26,28 +25,34 @@
       TestFactoryServiceClient factoryServiceClient,
-      TestWorkspaceServiceClient workspaceServiceClient) {
+      TestWorkspaceServiceClient workspaceServiceClient,
+      LoginPage loginPage) {
```
You moved TestFactory into the test module. Unfortunately we can't mix page objects and selenium core part. 
Describe usecase.

tolusha commented on selenium/che-selenium-test/src/test/resources/conf/selenium.properties:
```diff
@@ -24,3 +24,17 @@ github.auxiliary.password=
 google.password=
+
+# Define configuration for che default test user.
+# Note that for single user assembly the configuration should be:
```
It is n't clear if use has to be predefined in single user assembly. Could you clarify pls

akorneta replied:
added default values, so for single user che it's not needed to provide default user data

tolusha commented on selenium/che-selenium-core/src/main/java/org/eclipse/che/selenium/core/requestfactory/TestUserHttpJsonRequestFactory.java:
```diff
@@ -10,22 +10,35 @@
   protected String getAuthToken() {
-    return this.authToken;
+    try {
+      return authServiceClient.login(email, password);
```
Can we login once and then return authentication token?

dmytro-ndp replied:
We need to update token regularly because of little default token timeout around 5 minutes.

tolusha commented on selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/core/CheSeleniumMultiUserModule.java:
```diff
@@ -0,0 +1,50 @@
+    bind(TestUserServiceClient.class).to(CheTestUserServiceClient.class);
+    bind(HttpJsonRequestFactory.class).to(TestUserHttpJsonRequestFactory.class);
+    bind(TestUserHttpJsonRequestFactory.class).to(TestCheDefaultUserHttpJsonRequestFactory.class);
+    bind(TestMachineServiceClient.class).to(CheTestMachineServiceClient.class);
```
There is mess with class names.
`CheTest`, `Test`, `TestChe`

dmytro-ndp replied:
Let me explain to Anton the naming scheme:
1. Usually we add **Test** prefix to the classes which are specific to selenium tests like the following:
> AuthServiceClient -> **Test**AuthServiceClient
> HttpJsonRequestFactory -> **Test**UserHttpJsonRequestFactory
2. In case of different implementation for Che, Codenvy Onprem, Keycloak, Codenvy SaaS, we add name of product at start of class name:
> TestAuthServiceClient -> **Che**TestAuthServiceClient
or -> **Onpem**TestAuthServiceClient
or -> **Keycloak**TestAuthServiceClient

According to this scheme, correct name of class 
> Test**Che**DefaultUserHttpJsonRequestFactory > **Che**TestDefaultUserHttpJsonRequestFactory.

akorneta replied:
ok I'll fix that

tolusha commented on selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/core/CheSeleniumSuiteModule.java:
```diff
@@ -78,44 +66,33 @@ public void configure() {
-            .build(TestUserFactory.class));
-
-    bind(AdminTestUser.class).to(CheAdminTestUser.class);
+    if (parseBoolean(System.getenv(CHE_MULTIUSER_VARIABLE))) {
```
I can see the same bindings in both CheSeleniumMultiUserModule and CheSeleniumSingleUserModule.
Let's keep them in CheSeleniumSuiteModule

dmytro-ndp commented on selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/core/factory/TestFactoryInitializer.java:
```diff
@@ -80,7 +80,7 @@ public TestFactory fromUrl(String url) throws Exception {
         requestFactory.fromUrl(apiEndpointProvider.get() + "factory/resolver");
     httpJsonRequest.setBody(singletonMap("url", url));
-    HttpJsonResponse response = httpJsonRequest.request();
+    HttpJsonResponse respoTestCheDefaultUserHttpJsonRequestFactorynse = httpJsonRequest.request();
```
type

dmytro-ndp commented:
@akorneta: thank you for quick fixup.
I would appropriate it if commit messages will reflect content of changes.

tolusha commented:
There are 2 requirements to merge into master branch:
* don't mix page objects with core part of selenium tests
* ensure running tests for codenvy/saas projects

tolusha commented:
It is ok to be merged into multiuser branch and unblock platform team workking

skabashnyuk commented:
@tolusha  we do our best to met your requirements before merge che-multiuser branch in master
